ProjectEureka/project-eureka-spring
# PE-32| Created CRUD Operations for Answers table.

### REQUIREMENTS FOR MERGE

- **GET** response works for an empty database, database with your entries, and by specific id
- **POST** operation works by seeing new entries on postman and mongo
- **PUT** operation works by properly updating the specific entry on both mongo and postman
- **DELETE** operation works by deleting all previously posted entries in the database.

**AFTER YOUR PULL REQUEST, THE DATABASE MUST BE EMPTY**



### SET UP ENVIORNMENT
1. If you have not already, clone the repository **project-eureka-spring** and download Postman for desktop
2. In your editor/terminal, checkout in to the branch "PE-32-AnswersTableRestAPI" 

> **git checkout PE-32-AnswersTableRestAPI**

3. Navigate to src/main/resources. You will have to enter the correct credentials for the "spring.data.mongodb.uri" 
![Screen Shot 2021-01-31 at 5 31 29 PM](https://user-images.githubusercontent.com/48227740/106405175-2a69b580-63ea-11eb-868f-f834a9fa2317.png)
replace with the proper credentials. They are posted in discord in the backend channel and the project google drive. Once replaced, click save.

4. In your terminal, make sure you are in the directory that contains the project-eureka-spring folder and start the sprinboot application via the command 

> **mvn spring-boot:run**
Verify it is running by seeing this specific log
![VerifySpringRun](https://user-images.githubusercontent.com/48227740/106403293-b7a90c00-63e2-11eb-9566-fd3ceccfec00.png)
### TEST CRUD OPERATIONS

1. Open Postman and select **GET** for the url paste in **http://localhost:8080/api/v1/answers**. Click send. The following response should be an empty body, signaling that the answers collection on Mongo is empty.
![Screen Shot 2021-01-31 at 4 45 22 PM](https://user-images.githubusercontent.com/48227740/106403463-b7f5d700-63e3-11eb-8ea1-cab21fbcf339.png)
![Screen Shot 2021-01-31 at 4 47 55 PM](https://user-images.githubusercontent.com/48227740/106403540-1327c980-63e4-11eb-8355-008718fef7e5.png)
2. Now Select **POST** and using the same url. In the window directly under the url, click body. Select the indicator "raw" and set the body type to "JSON"
![Screen Shot 2021-01-31 at 4 50 59 PM](https://user-images.githubusercontent.com/48227740/106403629-803b5f00-63e4-11eb-9942-f89c4dd97e86.png)
3. In the body of the window, copy the same style and fields to input data. When inputting data, select your own data for the fields
![PostBody](https://user-images.githubusercontent.com/48227740/106403677-b973cf00-63e4-11eb-856b-53eba86074b2.png)
 4. Click Send. Your following response should like this:
![PostmanPostResponse](https://user-images.githubusercontent.com/48227740/106403719-e0ca9c00-63e4-11eb-9b56-bc3915e35d50.png)
**If a response is given, this signals that your data was sent to mongo properly. Verify by logging into the mongo database and checking the collections for the Project Eureka database. There should be a table labeled "Answers" and your entry will be there.**
![MongoCheck](https://user-images.githubusercontent.com/48227740/106403792-2ab38200-63e5-11eb-8bce-27efd3e5eb29.png)
5. Repeat steps 3 and 4 to post another entry into the mongo database. If this is successful, then that will verify that the **POST** action works
6. Verify the **GET** by repeating step 1. However, the response body should have the entries you already have entered.
![GetRsponseWithMultipleEntries](https://user-images.githubusercontent.com/48227740/106403893-94cc2700-63e5-11eb-958d-2d0a88e4cab7.png)
7. While still on the **GET** url, add and "/" and the "id" of one of your entries. These ids can be copied from the postman body and pasted into the url. 
8. Click "Send" The following response should be the entry containing the id of the entry you put in the url. This will verify that the getByID is working properly
![GetByID](https://user-images.githubusercontent.com/48227740/106404108-76b2f680-63e6-11eb-9729-1a7f855b549b.png)
![getByIDResponse](https://user-images.githubusercontent.com/48227740/106404091-63079000-63e6-11eb-8db3-db808e672f60.png)
9. Keep the same url header and select **PUT**
![PutRequestHeader](https://user-images.githubusercontent.com/48227740/106404149-b1b52a00-63e6-11eb-8fed-5c59aa1efc5a.png)

10. In the body below the header, change the data of the fields
![PutBody](https://user-images.githubusercontent.com/48227740/106404174-cb567180-63e6-11eb-99df-51baca4589c8.png)
Click "Send" The following response should be an entry with the same id, but different fields reflecting the changes you made. 
![PutBodyResponse](https://user-images.githubusercontent.com/48227740/106404215-f93bb600-63e6-11eb-809a-3266787dfc36.png)

11. Check that the properly implemented by running a get request in postman and checking the entry in mongo. If the changes are correct, this will verify that **PUT** is properly working
12. Keeping the same url, change to **DELETE**
13. Click "Send". This will delete the entry with that specific id from the mongo database. Verify the delete on the mongo database. Additionally, your postman response should look like so: 
![DeleteResponse](https://user-images.githubusercontent.com/48227740/106404402-8aab2800-63e7-11eb-9dc6-d23b962070ba.png)
If you try to click delete with the same id, you should get the following response.
![GetAfterDelete](https://user-images.githubusercontent.com/48227740/106404418-9c8ccb00-63e7-11eb-9e1b-5074acecb01b.png)
This response will be the same if you try a **GET** request with the same header. 
14. Repeat the steps to delete all of your entries. Make sure to delete every entry and to change the id to those already within the database
15. If successful, the mongo database should have no entries. Running a **GET** request will have the same result as step 1. If this is done, this will verify the **DELETE** operation is working.
